atomics/include/desul/atomics/SYCL.hpp

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright (c) 2019, Lawrence Livermore National Security, LLC
+and DESUL project contributors. See the COPYRIGHT file for details.
+Source: https://github.com/desul/desul
+
+SPDX-License-Identifier: (BSD-3-Clause)
+*/
+#ifndef DESUL_ATOMICS_SYCL_HPP_
+#define DESUL_ATOMICS_SYCL_HPP_
+
+#ifdef DESUL_HAVE_SYCL_ATOMICS
+#include "desul/atomics/Common.hpp"
+
+namespace desul {
+namespace Impl {
+template<class T>
+struct is_sycl_atomic_type {
+  static constexpr bool value = std::is_same<T, int>::value ||
+                                std::is_same<T, unsigned int>::value ||
+				std::is_same<T, long>::value ||
+				std::is_same<T, unsigned long>::value ||
+				std::is_same<T, long long>::value ||
+                                std::is_same<T, unsigned long long int>::value ||
+				std::is_same<T, float>::value ||
+				std::is_same<T, double>::value;
+};
+} // Impl
+
+// Atomic Add
+template<class T, class MemoryOrder/*, class MemoryScope*/>
```

Comment:
I would prefer to use the `SYCL` versions for all `MemoryScope`s but that doesn't work at the moment because host versions lead to ambiguous calls. I would prefer fixing/improving this behavior in a follow-up pull request such that fallback options are only used if the specialization/backend doesn't provide an implementation.

Reply:
What do you mean SYCL versions? These should always take the desul version, or should *at least* take the desul versions.  Do we need to come up with a scheme to avoid ambiguity due to device and host versions appearing in the same translation unit?

Reply:
I meant to say that `SYCL`'s `atomic_ref` can also be used for wider memory scopes and I would think it makes sense to use it. Defining this here for all template arguments leads to ambiguous calls, though.